Login from new device emails not sending

It looks like it doesn't understand the timezone name `PT` which we're getting back from somewhere (IP address geocoding?). Should be an easy fix

```
4d2c8c89-e3bb-41d4-8500-74fe9818a6c5] 2021-04-13T13:41:15-07:00 ERROR metabase.models.login-history Error sending 'login from new device' notification email
clojure.lang.ExceptionInfo: Conversion failed {:path [[#object[java_time.graph.Types 0x3bc54985 "[java.lang.CharSequence]"] #object[java_time.graph.Types 0x3f498d04 "[java.time.ZoneId]"]]], :arguments ["PT"], :to java.time.ZoneId}
	at java_time.zone$zone_id$fn__25947.invoke(zone.clj:73)
	at java_time.zone$zone_id.invokeStatic(zone.clj:73)
	at java_time.zone$zone_id.invoke(zone.clj:73)
	at metabase.email.messages$send_login_from_new_device_email_BANG_.invokeStatic(messages.clj:193)
	at metabase.email.messages$send_login_from_new_device_email_BANG_.invoke(messages.clj:187)
	at metabase.models.login_history$post_insert$fn__57210.invoke(login_history.clj:71)
	at metabase.models.login_history$post_insert.invokeStatic(login_history.clj:69)
	at metabase.models.login_history$post_insert.invoke(login_history.clj:65)
	at toucan.models$fn__23565$fn__23676$G__23584__23681.invoke(models.clj:178)
	at toucan.db$insert_BANG_.invokeStatic(db.clj:611)
	at toucan.db$insert_BANG_.invoke(db.clj:597)
	at metabase.api.session$fn__73514$record_login_history_BANG___73519$fn__73520.invoke(session.clj:33)
	at metabase.api.session$fn__73514$record_login_history_BANG___73519.invoke(session.clj:31)
	at metabase.api.session$fn__73547$create_session_BANG___73539__73552$fn__73553.invoke(session.clj:61)
	at metabase.api.session$fn__73547$create_session_BANG___73539__73552.invoke(session.clj:49)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at metabase.api.session$fn__73574$create_session_BANG___73566__73579$fn__73580.invoke(session.clj:69)
	at metabase.api.session$fn__73574$create_session_BANG___73566__73579.invoke(session.clj:64)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at clojure.lang.MultiFn.invoke(MultiFn.java:239)
	at metabase.api.session$fn__73626$email_login__73631$fn__73632.invoke(session.clj:108)
	at metabase.api.session$fn__73626$email_login__73631.invoke(session.clj:103)
	at metabase.api.session$fn__73653$login__73658$fn__73659.invoke(session.clj:124)
	at metabase.api.session$fn__73653$login__73658.invoke(session.clj:118)
	at metabase.api.session$fn__73675$do_login__73678.invoke(session.clj:151)
	at metabase.api.session$fn__73675$fn__73681$fn__73682$fn__73683.invoke(session.clj:159)
	at throttle.core$do_with_throttling.invokeStatic(core.clj:93)
	at throttle.core$do_with_throttling.invoke(core.clj:87)
	at metabase.api.session$fn__73675$fn__73681$fn__73682.invoke(session.clj:157)
	at throttle.core$do_with_throttling.invokeStatic(core.clj:93)
	at throttle.core$do_with_throttling.invoke(core.clj:87)
	at metabase.api.session$fn__73675$fn__73681.invoke(session.clj:157)
	at metabase.api.session$do_http_400_on_error.invokeStatic(session.clj:134)
	at metabase.api.session$do_http_400_on_error.invoke(session.clj:132)
	at metabase.api.session$fn__73675.invokeStatic(session.clj:156)
	at metabase.api.session$fn__73675.invoke(session.clj:144)
	at compojure.core$wrap_response$fn__1996.invoke(core.clj:160)
	at compojure.core$wrap_route_middleware$fn__1980.invoke(core.clj:132)
	at compojure.core$wrap_route_info$fn__1985.invoke(core.clj:139)
	at compojure.core$wrap_route_matches$fn__1989.invoke(core.clj:151)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008.invoke(core.clj:200)
	at metabase.api.session$_PLUS_log_all_request_failures$fn__73805.invoke(session.clj:355)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008.invoke(core.clj:200)
	at compojure.core$make_context$handler__2036.invoke(core.clj:287)
	at compojure.core$make_context$fn__2038.invoke(core.clj:296)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$make_context$fn__2038.invoke(core.clj:297)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at metabase.api.routes$fn__75715$fn__75717.invoke(routes.clj:70)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008.invoke(core.clj:200)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.server.routes$fn__77986$fn__77987.doInvoke(routes.clj:57)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008.invoke(core.clj:200)
	at compojure.core$make_context$handler__2036.invoke(core.clj:287)
	at compojure.core$make_context$fn__2038.invoke(core.clj:296)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__1989.invoke(core.clj:153)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__1989.invoke(core.clj:153)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__1989.invoke(core.clj:153)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008$f__2009$respond_SINGLEQUOTE___2010.invoke(core.clj:197)
	at metabase.server.routes$fn__77974$fn__77976.invoke(routes.clj:41)
	at compojure.core$routes$fn__2008$f__2009.invoke(core.clj:198)
	at compojure.core$routes$fn__2008.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__75706.invoke(exceptions.clj:98)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__75703.invoke(exceptions.clj:86)
	at metabase.server.middleware.log$log_api_call$fn__77902$fn__77903.invoke(log.clj:195)
	at toucan.db$_do_with_call_counting.invokeStatic(db.clj:216)
	at toucan.db$_do_with_call_counting.invoke(db.clj:209)
	at metabase.server.middleware.log$log_api_call$fn__77902.invoke(log.clj:189)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__77538.invoke(browser_cookie.clj:30)
	at metabase.server.middleware.security$add_security_headers$fn__75668.invoke(security.clj:142)
	at metabase.server.middleware.json$wrap_json_body$fn__77683.invoke(json.clj:62)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__77701.invoke(json.clj:98)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__78243.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__78259.invoke(params.clj:69)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__35489.invoke(misc.clj:58)
	at metabase.server.middleware.session$bind_current_user$fn__42461$fn__42462.invoke(session.clj:248)
	at metabase.server.middleware.session$do_with_current_user.invokeStatic(session.clj:229)
	at metabase.server.middleware.session$do_with_current_user.invoke(session.clj:221)
	at metabase.server.middleware.session$bind_current_user$fn__42461.invoke(session.clj:247)
	at metabase.server.middleware.session$wrap_current_user_info$fn__42448.invoke(session.clj:207)
	at metabase.server.middleware.session$wrap_session_id$fn__42434.invoke(session.clj:153)
	at metabase.server.middleware.auth$wrap_api_key$fn__75601.invoke(auth.clj:27)
	at ring.middleware.cookies$wrap_cookies$fn__78163.invoke(cookies.clj:216)
	at metabase.server.middleware.misc$add_content_type$fn__35472.invoke(misc.clj:27)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__35497.invoke(misc.clj:75)
	at ring.middleware.gzip$wrap_gzip$fn__78205.invoke(gzip.clj:86)
	at metabase.server.middleware.misc$bind_request$fn__35500.invoke(misc.clj:92)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__77919.invoke(ssl.clj:48)
	at metabase.server$async_proxy_handler$fn__77482.invoke(server.clj:71)
	at metabase.server.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.time.zone.ZoneRulesException: Unknown time-zone ID: PT
	at java.base/java.time.zone.ZoneRulesProvider.getProvider(Unknown Source)
	at java.base/java.time.zone.ZoneRulesProvider.getRules(Unknown Source)
	at java.base/java.time.ZoneRegion.ofId(Unknown Source)
	at java.base/java.time.ZoneId.of(Unknown Source)
	at java.base/java.time.ZoneId.of(Unknown Source)
	at java_time.zone$fn__25929.invokeStatic(zone.clj:32)
	at java_time.zone$fn__25929.invoke(zone.clj:31)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at java_time.defconversion$combinations$iter__25048__25052$fn__25053$fn__25054$fn__25055.doInvoke(defconversion.clj:28)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at java_time.defconversion$wrap_validation$fn__25045.invoke(defconversion.clj:21)
	... 202 more
```

--- a/src/metabase/email/messages.clj
+++ b/src/metabase/email/messages.clj
@@ -189,10 +189,8 @@
   login history infomation as returned by `metabase.models.login-history/human-friendly-infos`."
   [{user-id :user_id, :keys [timestamp timezone], :as login-history}]
   (let [user-info    (db/select-one ['User [:first_name :first-name] :email :locale] :id user-id)
-        timestamp    (cond-> timestamp
-                       timezone (u.date/with-time-zone-same-instant (t/zone-id timezone))
-                       true     (u.date/format-human-readable (or (:locale user-info)
-                                                                  (i18n/site-locale))))
+        user-locale  (or (:locale user-info) (i18n/site-locale))
+        timestamp    (u.date/format-human-readable timestamp user-locale)
         context      (merge (common-context)
                             {:first-name (:first-name user-info)
                              :device     (:device_description login-history)

--- a/src/metabase/email/messages.clj
+++ b/src/metabase/email/messages.clj
@@ -187,7 +187,7 @@
 (defn send-login-from-new-device-email!
   "Format and send an email informing the user that this is the first time we've seen a login from this device. Expects
   login history infomation as returned by `metabase.models.login-history/human-friendly-infos`."
-  [{user-id :user_id, :keys [timestamp timezone], :as login-history}]
+  [{user-id :user_id, :keys [timestamp], :as login-history}]
   (let [user-info    (db/select-one ['User [:first_name :first-name] :email :locale] :id user-id)
         user-locale  (or (:locale user-info) (i18n/site-locale))
         timestamp    (u.date/format-human-readable timestamp user-locale)

--- a/src/metabase/models/login_history.clj
+++ b/src/metabase/models/login_history.clj
@@ -1,5 +1,6 @@
 (ns metabase.models.login-history
   (:require [clojure.tools.logging :as log]
+            [java-time :as t]
             [metabase.email.messages :as email.messages]
             [metabase.models.setting :refer [defsetting]]
             [metabase.server.request.util :as request.u]
@@ -30,7 +31,7 @@
                  :timezone (timezone-display-name timezone))
           (update :timestamp (fn [timestamp]
                                (if (and timestamp timezone)
-                                 (u.date/with-time-zone-same-instant timestamp timezone)
+                                 (t/zoned-date-time (u.date/with-time-zone-same-instant timestamp timezone) timezone)
                                  timestamp)))
           (update :device_description request.u/describe-user-agent)))))
 

--- a/src/metabase/util/date_2.clj
+++ b/src/metabase/util/date_2.clj
@@ -13,7 +13,7 @@
             [schema.core :as s])
   (:import [java.time DayOfWeek Duration Instant LocalDate LocalDateTime LocalTime OffsetDateTime OffsetTime Period
             ZonedDateTime]
-           [java.time.format DateTimeFormatter DateTimeFormatterBuilder FormatStyle]
+           [java.time.format DateTimeFormatter DateTimeFormatterBuilder FormatStyle TextStyle]
            [java.time.temporal Temporal TemporalAdjuster WeekFields]
            org.threeten.extra.PeriodDuration))
 
@@ -106,13 +106,22 @@
                     (.toFormatter builder))
    OffsetTime     (let [builder (doto (DateTimeFormatterBuilder.)
                                   (.append (DateTimeFormatter/ofLocalizedTime FormatStyle/MEDIUM))
-                                  (.appendPattern " (OOOO)"))]
+                                  (.appendLiteral " (")
+                                  (.appendLocalizedOffset TextStyle/FULL)
+                                  (.appendLiteral ")"))]
                     (.toFormatter builder))
    OffsetDateTime (let [builder (doto (DateTimeFormatterBuilder.)
                                   (.appendLocalized FormatStyle/LONG FormatStyle/MEDIUM)
-                                  (.appendPattern " (OOOO)"))]
+                                  (.appendLiteral " (")
+                                  (.appendLocalizedOffset TextStyle/FULL)
+                                  (.appendLiteral ")"))]
                     (.toFormatter builder))
-   ZonedDateTime  (DateTimeFormatter/ofLocalizedDateTime FormatStyle/LONG)})
+   ZonedDateTime  (let [builder (doto (DateTimeFormatterBuilder.)
+                                  (.appendLocalized FormatStyle/LONG FormatStyle/MEDIUM)
+                                  (.appendLiteral " (")
+                                  (.appendZoneText TextStyle/FULL)
+                                  (.appendLiteral ")"))]
+                    (.toFormatter builder))})
 
 (defn format-human-readable
   "Format a temporal value `t` in a human-friendly way for `locale` (by default, the current User's locale).

--- a/test/metabase/models/login_history_test.clj
+++ b/test/metabase/models/login_history_test.clj
@@ -2,10 +2,11 @@
   (:require [clojure.string :as str]
             [clojure.test :refer :all]
             [environ.core :as env]
-            [metabase.db :as mdb]
+            [java-time :as t]
             [metabase.models :refer [LoginHistory User]]
             [metabase.models.login-history :as login-history]
             [metabase.models.setting.cache :as setting.cache]
+            [metabase.server.request.util :as request.u]
             [metabase.test :as mt]
             [metabase.util.schema :as su]
             [schema.core :as s]))
@@ -34,51 +35,61 @@
                  (#'login-history/first-login-ever? history-1))))))))))
 
 (deftest send-email-on-first-login-from-new-device-test
-  (mt/with-temp User [{user-id :id, email :email, first-name :first_name}]
-    (let [device (str (java.util.UUID/randomUUID))]
-      (testing "send email on first login from *new* device (but not first login ever)"
-        (mt/with-fake-inbox
-          (mt/with-temp* [LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]
-                          LoginHistory [_ {:user_id user-id, :device_id device, :timestamp #t "2021-04-02T15:52:00-07:00[US/Pacific]"}]]
-            (is (schema= {(s/eq email)
-                          [{:from    su/Email
-                            :to      (s/eq [email])
-                            :subject (s/eq (format "We've Noticed a New Metabase Login, %s" first-name))
-                            :body    [(s/one {:type    (s/eq "text/html; charset=utf-8")
-                                              :content s/Str}
-                                             "HTML body")]}]}
-                         @mt/inbox))
-            (let [message (-> @mt/inbox (get email) first :body first :content)]
-              (testing (format "\nMessage = %s" (pr-str message))
-                (is (string? message))
-                (when (string? message)
-                  (is (str/includes? message "We've noticed a new login on your Metabase account."))
-                  (is (str/includes? message "We noticed a login on your Metabase account from a new device."))
-                  (is (str/includes? message "Browser (Chrome/Windows) - Unknown location"))
-                  (if (= (mdb/db-type) :h2)
-                    (str/includes? message "April 2 3:52 PM (GMT-07:00)")
-                    (str/includes? message "April 2 10:52 PM (GMT)")))))
+  (testing "User should get an email the first time they log in from a new device (#14313, #15603)"
+    (mt/with-temp User [{user-id :id, email :email, first-name :first_name}]
+      (let [device (str (java.util.UUID/randomUUID))]
+        (testing "send email on first login from *new* device (but not first login ever)"
+          (mt/with-fake-inbox
+            ;; mock out the IP address geocoding function so we can make sure it handles timezones like PST correctly
+            ;; (#15603)
+            (with-redefs [request.u/geocode-ip-addresses (fn [ip-addresses]
+                                                           (into {} (for [ip-address ip-addresses]
+                                                                      [ip-address
+                                                                       {:description "San Francisco, California, United States"
+                                                                        :timezone    (t/zone-id "America/Los_Angeles")}])))]
+              (mt/with-temp* [LoginHistory [_ {:user_id   user-id
+                                               :device_id (str (java.util.UUID/randomUUID))}]
+                              LoginHistory [_ {:user_id   user-id
+                                               :device_id device
+                                               :timestamp #t "2021-04-02T15:52:00-07:00[US/Pacific]"}]]
 
-            (testing "don't send email on subsequent login from same device"
-              (mt/reset-inbox!)
-              (mt/with-temp LoginHistory [_ {:user_id user-id, :device_id device}]
-                (is (= {}
-                       @mt/inbox)))))))))
+                (is (schema= {(s/eq email)
+                              [{:from    su/Email
+                                :to      (s/eq [email])
+                                :subject (s/eq (format "We've Noticed a New Metabase Login, %s" first-name))
+                                :body    [(s/one {:type    (s/eq "text/html; charset=utf-8")
+                                                  :content s/Str}
+                                                 "HTML body")]}]}
+                             @mt/inbox))
+                (let [message (-> @mt/inbox (get email) first :body first :content)]
+                  (testing (format "\nMessage = %s" (pr-str message))
+                    (is (string? message))
+                    (when (string? message)
+                      (doseq [expected-str ["We've noticed a new login on your Metabase account."
+                                            "We noticed a login on your Metabase account from a new device."
+                                            "Browser (Chrome/Windows) - San Francisco, California, United States"
+                                            "April 2, 2021 3:52:00 PM (Pacific Daylight Time)"]]
+                        (is (str/includes? message expected-str))))))
+
+                (testing "don't send email on subsequent login from same device"
+                  (mt/reset-inbox!)
+                  (mt/with-temp LoginHistory [_ {:user_id user-id, :device_id device}]
+                    (is (= {}
+                           @mt/inbox)))))))))))
 
   (testing "don't send email if the setting is disabled by setting MB_SEND_EMAIL_ON_FIRST_LOGIN_FROM_NEW_DEVICE=FALSE"
     (mt/with-temp User [{user-id :id, email :email, first-name :first_name}]
-      (testing "send email on first login from new device"
-        (try
-          (mt/with-fake-inbox
-            ;; can't use `mt/with-temporary-setting-values` here because it's a read-only setting
-            (with-redefs [env/env (assoc env/env :mb-send-email-on-first-login-from-new-device "FALSE")]
-              ;; flush the Setting cache so it picks up the env var value for the `send-email-on-first-login-from-new-device` setting
-              (setting.cache/restore-cache!)
-              (mt/with-temp* [LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]
-                              LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]]
-                (is (= {}
-                       @mt/inbox)))))
-          (finally
-            ;; flush the cache again so the original value of `send-email-on-first-login-from-new-device` gets
-            ;; restored
-            (setting.cache/restore-cache!)))))))
+      (try
+        (mt/with-fake-inbox
+          ;; can't use `mt/with-temporary-setting-values` here because it's a read-only setting
+          (with-redefs [env/env (assoc env/env :mb-send-email-on-first-login-from-new-device "FALSE")]
+            ;; flush the Setting cache so it picks up the env var value for the `send-email-on-first-login-from-new-device` setting
+            (setting.cache/restore-cache!)
+            (mt/with-temp* [LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]
+                            LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]]
+              (is (= {}
+                     @mt/inbox)))))
+        (finally
+          ;; flush the cache again so the original value of `send-email-on-first-login-from-new-device` gets
+          ;; restored
+          (setting.cache/restore-cache!))))))

--- a/test/metabase/models/login_history_test.clj
+++ b/test/metabase/models/login_history_test.clj
@@ -8,6 +8,7 @@
             [metabase.models.setting.cache :as setting.cache]
             [metabase.server.request.util :as request.u]
             [metabase.test :as mt]
+            [metabase.util.date-2 :as u.date]
             [metabase.util.schema :as su]
             [schema.core :as s]))
 
@@ -68,7 +69,8 @@
                       (doseq [expected-str ["We've noticed a new login on your Metabase account."
                                             "We noticed a login on your Metabase account from a new device."
                                             "Browser (Chrome/Windows) - San Francisco, California, United States"
-                                            "April 2, 2021 3:52:00 PM (Pacific Daylight Time)"]]
+                                            ;; `format-human-readable` has slightly different output on different JVMs
+                                            (u.date/format-human-readable #t "2021-04-02T15:52:00-07:00[US/Pacific]")]]
                         (is (str/includes? message expected-str))))))
 
                 (testing "don't send email on subsequent login from same device"

--- a/test/metabase/util/date_2_test.clj
+++ b/test/metabase/util/date_2_test.clj
@@ -158,11 +158,11 @@
   ;; strings are localized slightly differently on different JVMs. For places where there are multiple possible
   ;; correct results, we'll use a set with all the possibilities below and check membership
   (doseq [[t expected] {#t "2021-04-02T14:42:09.524392-07:00[US/Pacific]" ; ZonedDateTime
-                        {:en-US #{"April 2, 2021 2:42:09 PM PDT"
-                                  "April 2, 2021 at 2:42:09 PM PDT"}
-                         :es-MX #{"2 de abril de 2021 02:42:09 PM PDT"
-                                  "2 de abril de 2021, 14:42:09 PDT"
-                                  "2 de abril de 2021 a las 14:42:09 PDT"}}
+                        {:en-US #{"April 2, 2021 2:42:09 PM (Pacific Daylight Time)"
+                                  "April 2, 2021 at 2:42:09 PM (Pacific Daylight Time)"}
+                         :es-MX #{"2 de abril de 2021 02:42:09 PM (Hora de verano del Pacífico)"
+                                  "2 de abril de 2021, 14:42:09 (Hora de verano del Pacífico)"
+                                  "2 de abril de 2021 a las 14:42:09 (Hora de verano del Pacífico)"}}
 
                         #t "2021-04-02T14:42:09.524392-07:00" ; OffsetDateTime
                         {:en-US #{"April 2, 2021 2:42:09 PM (GMT-07:00)"

--- a/test/metabase/util/date_2_test.clj
+++ b/test/metabase/util/date_2_test.clj
@@ -159,10 +159,11 @@
   ;; correct results, we'll use a set with all the possibilities below and check membership
   (doseq [[t expected] {#t "2021-04-02T14:42:09.524392-07:00[US/Pacific]" ; ZonedDateTime
                         {:en-US #{"April 2, 2021 2:42:09 PM (Pacific Daylight Time)"
-                                  "April 2, 2021 at 2:42:09 PM (Pacific Daylight Time)"}
+                                  "April 2, 2021, 2:42:09 PM (Pacific Daylight Time)"}
                          :es-MX #{"2 de abril de 2021 02:42:09 PM (Hora de verano del Pacífico)"
-                                  "2 de abril de 2021, 14:42:09 (Hora de verano del Pacífico)"
-                                  "2 de abril de 2021 a las 14:42:09 (Hora de verano del Pacífico)"}}
+                                  "2 de abril de 2021 14:42:09 (Hora de verano del Pacífico)"
+                                  "2 de abril de 2021 14:42:09 (hora de verano del Pacífico)"
+                                  "2 de abril de 2021, 14:42:09 (Hora de verano del Pacífico)"}}
 
                         #t "2021-04-02T14:42:09.524392-07:00" ; OffsetDateTime
                         {:en-US #{"April 2, 2021 2:42:09 PM (GMT-07:00)"

